### ISSUE TITLE
Research: Research best way for a user to set an app component to loading/blocked state

### ISSUE THREAD

**Issue by driskull** [Esri/calcite-app-components#161]
### Background

Research best way for a user to set an app component to` loading`/`disabled`/`blocked` state

#### Use Case

User clicks on an action within a panel or a flow which fires off some asynchronous task. During the task, we want to display to the user that something is happening and disable further navigation until the task is complete in order to not mess up the task from completing and showing a result.

### Desired Outcome

- Any Component can be independently "blocked", "loading", "disabled".

We have a documented approach to how a user can handle the above scenario.

### Ideas

TODO

**Comment by jcfranco** [comment]
Another option could be to introduce a component/utility similar to the 3.x's [`busyIndicator`](https://developers.arcgis.com/javascript/3/jsapi/esri.dijit.util.busyindicator-amd.html), which takes care of blocking a target element by placing an overlay, and loading indicator until disabled.

**Comment by driskull** [comment]
> Another option could be to introduce a component/utility similar to the 3.x's busyIndicator

yeah, I think this would be the busy indicator: https://github.com/Esri/calcite-components/tree/master/src/components/calcite-progress

**Comment by jcfranco** [comment]
It's not the same though. `calcite-progress` just shows the indicator, whereas the util has some extra features to easily block different areas on a page.

**Comment by driskull** [comment]
Yep, we could use a utility component to block areas. Or maybe just add a class on the `calcite-progress` to adjust styling and set pointer events to none?

**Comment by jcfranco** [comment]
> Or maybe just add a class on the calcite-progress to adjust styling and set pointer events to none?

I think we'd need something more flexible, no? There's only so much a class can do.

**Comment by driskull** [comment]
> I think we'd need something more flexible, no? There's only so much a class can do.

Maybe. I was just thinking of a simple solution.

- It could be its own `CalciteState` component that could wrap anything and apply a loader or block things.
- It could be a utility that we have on each component we want to expose to it.

**Comment by pr3tori4n** [comment]
@driskull @jcfranco is there a simple way to see what a user sees today while waiting for things to fully load? You have any screenshots or a demo page you can share?

**Comment by driskull** [comment]
@kat10140 and @asangma  can you help with the requirements and then @pr3tori4n can work on an API design before any coding work.

**Comment by asangma** [comment]
TLDR: At the moment, we only need to have a property to use for TBD styling. 

Styling will need to adapt to each component and the moment for the user (loading vs processing), so I think step one would be to have a property that can be used to add styles on a per-component basis.

**Comment by asangma** [comment]
@jcfranco 
> There's only so much a class can do.

Agreed. I think a property or class is a good start though.
Also, the main loader states we have are 
- an animated graphic (hopefully doable via CSS)
- a loader bar like we have on the LayerList.
- skeleton screen type animation (low priority)

The skeleton screen would be the most complex one, and based on research we did, it only offers marginal improvements to the experience of waiting.

**Comment by pr3tori4n** [comment]
List of Components that need loading props:
* action
* block
* flow
* flow-item
* shell-panel
* pickList
* valueList

**Comment by pr3tori4n** [comment]
Not doing skeleton screens. Sticking to spinners/loading bars. Will be component specific designs.

**Comment by pr3tori4n** [comment]
@driskull we have a plan for loading for the near future. Do we also need a separate plan for blocked and disabled?

**Comment by driskull** [comment]
Seems like we would need a plan for `disabled` but I'm not sure we need anything for `blocked`. Wouldn't blocked be either disabled or loading? Need to define that use case.

**Comment by pr3tori4n** [comment]
@driskull can we split the conversation around disabled into another issue?

**Comment by driskull** [comment]
Sure

**Comment by pr3tori4n** [comment]
@asangma @kat10140 Can we consider this item done? We have an issue for implementing loading state.

**Comment by driskull** [comment]
Considering this as done. 

@asangma @kat10140 Can you create separate issues for adding `disabled` to the components that need it?
